Stack/Opc.Ua.Core/Security/Certificates/ReentrantSemaphoreSlim.cs

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 1996-2022 The OPC Foundation. All rights reserved.
+   The source code in this file is covered under a dual-license scenario:
+     - RCL: for OPC Foundation Corporate Members in good-standing
+     - GPL V2: everybody else
+   RCL license terms accompanied with this source code. See http://opcfoundation.org/License/RCL/1.00/
+   GNU General Public License as published by the Free Software Foundation;
+   version 2 of the License are accompanied with this source code. See http://opcfoundation.org/License/GPLv2
+   This source code is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+*/
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Opc.Ua
+{
+    /// <summary>
+    /// Defines a reentrant SemaphoreSlim Lock.
+    /// </summary>
+    internal class ReentrantSemaphoreSlim
+    {
+        #region Constructors
+        public ReentrantSemaphoreSlim()
+        {
+            m_semaphore = new SemaphoreSlim(1, 1);
+            m_reentrantCounter = 0;
+        }
+
+        public ReentrantSemaphoreSlim(int initialCount, int maxCount)
+        {
+            m_semaphore = new SemaphoreSlim(initialCount, maxCount);
+            m_reentrantCounter = 0;
+        }
+        #endregion
+
+        #region Public Members
+        /// <summary>
+        /// Wait method for synchronous usage
+        /// </summary>
+        public void Wait()
+        {
+            int counter = m_reentrantCounter;
```

Comment:
is this really working? It means a second thread can just execute ignoring the semaphore?

Reply:
I think you need to check the owning thread variable first, otherwise the semaphore functionality is undermined